lib/puppet/http/service/compiler.rb

```diff
@@ -129,6 +129,70 @@ def post_catalog(name, facts:, environment:, configured_environment: nil, transa
     [response, deserialize(response, Puppet::Resource::Catalog)]
   end
 
+  # TODO; document this API once it is finalized
+  def post_catalog4(certname, persistence, environment, facts: nil, trusted_facts: nil, transaction_uuid: nil, job_id: nil, options: nil)
+    body = {
+      certname: certname,
+      persistence: persistence,
+      environment: environment,
+      transaction_uuid: transaction_uuid,
+      job_id: job_id,
+      options: options
+    }
+    body.merge!({ facts: { values: facts }}) unless facts.nil?
+    body.merge!({ trusted_facts: { values: trusted_facts }}) unless trusted_facts.nil?
+    headers = {
+      'accept-encoding' => Puppet::Network::HTTP::Compression::ACCEPT_ENCODING,
+      'Content-Type' => 'text/json',
+      'Accept' => 'application/json',
+      Puppet::Network::HTTP::HEADER_PUPPET_VERSION => Puppet.version
+    }
+
+    # CODEREVIEW: The base url has v3 baked in and does not appear to be easily configured here
+    # Is this somewhat hacky work around acceptable?
+    v3_regex = /\/v3$/
+    url = with_base_url("")
+    if url.path =~ v3_regex
+      url.path = url.path.gsub(v3_regex, '') + '/v4/catalog'
+    else
+      #TODO: raise error
+    end
```

Comment:
There should be a `@url` instance variable (from the superclass) that you can use to extract the `host` and `port` for the currently configured endpoint (based on `server_list`, `server`, etc setttings). Then you can do something like:

```ruby
url = URI::HTTPS.build(host: @url.server, port: @url.port, path: Puppet::Util.uri_encode("/puppet/v4/catalog/#{name}")
```